components/raftstore-v2/src/operation/query/mod.rs

```diff
@@ -71,6 +74,151 @@ impl<'a, EK: KvEngine, ER: RaftEngine, T: raftstore::store::Transport>
+            current_version = self
+                .fsm
+                .peer()
+                .last_region_buckets()
+                .as_ref()
+                .map(|b| b.meta.version)
+                .unwrap_or_default();
+        }
+        let mut region_buckets: BucketStat;
+        if let (Some(bucket_ranges), Some(peer_region_buckets)) =
+            (bucket_ranges, self.fsm.peer().region_buckets())
+        {
+            assert_eq!(buckets.len(), bucket_ranges.len());
+            let mut i = 0;
+            region_buckets = peer_region_buckets.clone();
+            let mut meta = (*region_buckets.meta).clone();
+            if !buckets.is_empty() {
+                meta.version = gen_bucket_version(self.fsm.peer().term(), current_version);
+            }
+            meta.region_epoch = region_epoch;
+            for (bucket, bucket_range) in buckets.into_iter().zip(bucket_ranges) {
+                while i < meta.keys.len() && meta.keys[i] != bucket_range.0 {
+                    i += 1;
+                }
+                assert!(i != meta.keys.len());
+                // the bucket size is small and does not have split keys,
+                // then it should be merged with its left neighbor
+                let region_bucket_merge_size = self
+                    .store_ctx
+                    .coprocessor_host
+                    .cfg
+                    .region_bucket_merge_size_ratio
+                    * (self.store_ctx.coprocessor_host.cfg.region_bucket_size.0 as f64);
+                if bucket.keys.is_empty() && bucket.size <= (region_bucket_merge_size as u64) {
+                    meta.sizes[i] = bucket.size;
+                    // i is not the last entry (which is end key)
+                    assert!(i < meta.keys.len() - 1);
+                    // the region has more than one bucket
+                    // and the left neighbor + current bucket size is not very big
+                    if meta.keys.len() > 2
+                        && i != 0
+                        && meta.sizes[i - 1] + bucket.size
+                            < self.store_ctx.coprocessor_host.cfg.region_bucket_size.0 * 2
+                    {
+                        // bucket is too small
+                        region_buckets.left_merge(i);
+                        meta.left_merge(i);
+                        continue;
+                    }
+                } else {
+                    // update size
+                    meta.sizes[i] = bucket.size / (bucket.keys.len() + 1) as u64;
+                    // insert new bucket keys (split the original bucket)
+                    for bucket_key in bucket.keys {
+                        i += 1;
+                        region_buckets.split(i);
+                        meta.split(i, bucket_key);
+                    }
+                }
+                i += 1;
+            }
+            region_buckets.meta = Arc::new(meta);
+        } else {
+            assert_eq!(buckets.len(), 1);
```

Comment:
What if the condition at L136 is (Some. None)?

Reply:
In the beginning, the bucket range is always none,  so it should always generated one buckets. Otherwise, the buckets number is same as the given bucket ranges .

Reply:
I mean, what if `self.region_buckets` is reset (https://github.com/tikv/tikv/pull/14044#discussion_r1103959084), and an old `PeerMsg` has `buckets.len() > 1`?

Reply:
It will skip this refresh event if epoch is [stale].(https://github.com/tikv/tikv/blob/19fdab7f543a3ab355ae2cb94e3534231fee2438/components/raftstore-v2/src/operation/bucket.rs#L51-L59)

Reply:
You should comment that "`reset_region_buckets` is only called during split, refresh requests before split will be rejected due to stale epoch, so there won't be race between `reset_region_buckets` and the requests before it."

Reply:
Ping this one.

Reply:
Ping this one.

Reply:
yes, I have added some comment on this condition.